jormungandr/src/explorer/graphql/mod.rs

```diff
@@ -19,39 +19,391 @@ use super::indexing::{
+    ) -> FieldResult<Option<BlockConnection>> {
+        let epoch_data = match context.db.get_epoch(epoch.try_into()?).await {
+            Some(epoch_data) => epoch_data,
+            None => return Ok(None),
+        };
+
+        let epoch_lower_bound = self
+            .state
+            .state()
+            .blocks
+            .lookup(&epoch_data.first_block)
+            .map(|block| u32::from(block.chain_length))
+            .expect("Epoch lower bound");
+
+        let epoch_upper_bound = self
+            .state
+            .state()
+            .blocks
+            .lookup(&epoch_data.last_block)
+            .map(|block| u32::from(block.chain_length))
+            .expect("Epoch upper bound");
+
+        let boundaries = PaginationInterval::Inclusive(InclusivePaginationInterval {
+            lower_bound: 0,
+            upper_bound: epoch_upper_bound
+                .checked_sub(epoch_lower_bound)
+                .expect("pagination upper_bound to be greater or equal than lower_bound"),
+        });
+
+        let pagination_arguments = PaginationArguments {
+            first,
+            last,
+            before: before.map(u32::try_from).transpose()?,
+            after: after.map(u32::try_from).transpose()?,
+        }
+        .validate()?;
+
+        BlockConnection::new_async(boundaries, pagination_arguments, |range| async {
+            match range {
+                PaginationInterval::Empty => unreachable!("No blocks found (not even genesis)"),
+                PaginationInterval::Inclusive(range) => self
+                    .state
+                    .state()
+                    .get_block_hash_range(
+                        (range.lower_bound + epoch_lower_bound).into(),
+                        (range.upper_bound + epoch_lower_bound + 1).into(),
+                    )
+                    .iter()
+                    .map(|(hash, index)| (*hash, u32::from(*index) - epoch_lower_bound))
+                    .collect(),
+            }
+        })
+        .await
+        .map(Some)
+    }
+}
+
 pub struct Block {
     hash: HeaderHash,
+    contents: tokio::sync::Mutex<Option<Arc<ExplorerBlock>>>,
 }
 
 impl Block {
     async fn from_string_hash(hash: String, db: &ExplorerDB) -> FieldResult<Block> {
         let hash = HeaderHash::from_str(&hash)?;
-        let block = Block { hash };
+        let block = Block {
+            hash,
+            contents: Default::default(),
+        };
 
         block.get_explorer_block(db).await.map(|_| block)
     }
 
     fn from_valid_hash(hash: HeaderHash) -> Block {
-        Block { hash }
+        Block {
+            hash,
+            contents: Default::default(),
+        }
+    }
+
+    fn from_contents(block: Arc<ExplorerBlock>) -> Block {
+        Block {
+            hash: block.id(),
+            contents: tokio::sync::Mutex::new(Some(block)),
+        }
     }
 
     async fn get_explorer_block(&self, db: &ExplorerDB) -> FieldResult<Arc<ExplorerBlock>> {
```

Comment:
Could rename to `fetch_explorer_block` to reflect that it caches the retrieved block now.